Issues Running MariaDb4j on Glassfish

Hi,

I have a Spring MVC-based java (non-spring boot) application that is running on Glassfish 4.1.1. We use Java 8 for our development. When I simply reference the maven dependency in my POM file even without updating any of my code base to use MariaDb4j, during WAR deployment, I get the following error

`[2017-08-01 02:20:04,632] Artifact my_application:war: java.io.IOException: com.sun.enterprise.admin.remote.RemoteFailureException: Error occurred during deployment: Exception while deploying the app [my_application] : The lifecycle method [stop] must not throw a checked exception. Related annotation information: annotation [@javax.annotation.PreDestroy()] on annotated element [public void ch.vorburger.mariadb4j.MariaDB4jService.stop() throws ch.vorburger.exec.ManagedProcessException] of type [METHOD]. Please see server.log for more details.`

Following is the stack trace

```
java.lang.IllegalStateException: The lifecycle method [stop] must not throw a checked exception. Related annotation information: annotation [@javax.annotation.PreDestroy()] on annotated element [public void ch.vorburger.mariadb4j.MariaDB4jService.stop() throws ch.vorburger.exec.ManagedProcessException] of type [METHOD]
	at com.sun.enterprise.deployment.archivist.Archivist.readAnnotations(Archivist.java:519)
	at com.sun.enterprise.deployment.archivist.Archivist.readAnnotations(Archivist.java:447)
	at org.glassfish.web.deployment.archivist.WebArchivist.postAnnotationProcess(WebArchivist.java:338)
	at org.glassfish.web.deployment.archivist.WebArchivist.postAnnotationProcess(WebArchivist.java:91)
	at com.sun.enterprise.deployment.archivist.Archivist.readRestDeploymentDescriptors(Archivist.java:421)
	at com.sun.enterprise.deployment.archivist.Archivist.readDeploymentDescriptors(Archivist.java:396)
	at com.sun.enterprise.deployment.archivist.Archivist.open(Archivist.java:271)
	at com.sun.enterprise.deployment.archivist.Archivist.open(Archivist.java:280)
	at com.sun.enterprise.deployment.archivist.Archivist.open(Archivist.java:241)
	at com.sun.enterprise.deployment.archivist.ApplicationFactory.openArchive(ApplicationFactory.java:161)
	at org.glassfish.javaee.core.deployment.DolProvider.processDOL(DolProvider.java:203)
	at org.glassfish.javaee.core.deployment.DolProvider.load(DolProvider.java:227)
	at org.glassfish.javaee.core.deployment.DolProvider.load(DolProvider.java:96)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.loadDeployer(ApplicationLifecycle.java:881)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.setupContainerInfos(ApplicationLifecycle.java:821)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:377)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:219)
	at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:491)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:539)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:535)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:534)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:565)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:557)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:556)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1464)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.access$1300(CommandRunnerImpl.java:109)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1846)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1722)
	at com.sun.enterprise.v3.admin.AdminAdapter.doCommand(AdminAdapter.java:534)
	at com.sun.enterprise.v3.admin.AdminAdapter.onMissingResource(AdminAdapter.java:224)
	at org.glassfish.grizzly.http.server.StaticHttpHandlerBase.service(StaticHttpHandlerBase.java:189)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:459)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:167)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:206)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:180)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:283)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:200)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:132)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:111)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:536)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:591)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:571)
	at java.lang.Thread.run(Thread.java:745)
Caused by: The lifecycle method [stop] must not throw a checked exception. Related annotation information: annotation [@javax.annotation.PreDestroy()] on annotated element [public void ch.vorburger.mariadb4j.MariaDB4jService.stop() throws ch.vorburger.exec.ManagedProcessException] of type [METHOD]
	at org.glassfish.apf.impl.AnnotationProcessorImpl.process(AnnotationProcessorImpl.java:373)
	at org.glassfish.apf.impl.AnnotationProcessorImpl.process(AnnotationProcessorImpl.java:381)
	at org.glassfish.apf.impl.AnnotationProcessorImpl.processAnnotations(AnnotationProcessorImpl.java:289)
	at org.glassfish.apf.impl.AnnotationProcessorImpl.process(AnnotationProcessorImpl.java:217)
	at org.glassfish.apf.impl.AnnotationProcessorImpl.process(AnnotationProcessorImpl.java:134)
	at com.sun.enterprise.deployment.archivist.Archivist.processAnnotations(Archivist.java:627)
	at com.sun.enterprise.deployment.archivist.Archivist.readAnnotations(Archivist.java:463)
	... 53 more
Caused by: java.lang.IllegalArgumentException: The lifecycle method [stop] must not throw a checked exception
	at com.sun.enterprise.deployment.annotation.handlers.AbstractResourceHandler.validateAnnotatedLifecycleMethod(AbstractResourceHandler.java:186)
	at com.sun.enterprise.deployment.annotation.handlers.PreDestroyHandler.processAnnotation(PreDestroyHandler.java:69)
	at com.sun.enterprise.deployment.annotation.handlers.AbstractResourceHandler.processAnnotation(AbstractResourceHandler.java:142)
	at com.sun.enterprise.deployment.annotation.factory.SJSASFactory$LazyAnnotationHandler.processAnnotation(SJSASFactory.java:148)
	at org.glassfish.apf.impl.AnnotationProcessorImpl.process(AnnotationProcessorImpl.java:350)
	... 59 more
]]
```


I am wondering whether this library only works for Spring boot application. Has this library been ever tested on application deployed in glassfish?

The application is successfully deployed on Tomcat 8.5.16 but I have a need to make it run on GlassFish 4.1.x

--- a/mariaDB4j-core/src/main/java/ch/vorburger/mariadb4j/ManagedProcessRuntimeException.java
+++ b/mariaDB4j-core/src/main/java/ch/vorburger/mariadb4j/ManagedProcessRuntimeException.java
@@ -1,0 +1,39 @@
+/*
+ * #%L
+ * MariaDB4j
+ * %%
+ * Copyright (C) 2012 - 2018 Michael Vorburger
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package ch.vorburger.mariadb4j;
+
+import ch.vorburger.exec.ManagedProcess;
+import ch.vorburger.exec.ManagedProcessException;
+
+/**
+ * Exception thrown when unexpected thing happen in {@link ManagedProcess} and
+ * <a href="https://github.com/vorburger/MariaDB4j/issues/88">caller needs an unchecked instead of a checked exception</a>.
+ * 
+ * @author Michael Vorburger.ch
+ */
+public class ManagedProcessRuntimeException extends RuntimeException {
+
+    private static final long serialVersionUID = -415492090964791430L;
+
+    public ManagedProcessRuntimeException(ManagedProcessException original) {
+        super(original.getMessage(), original.getCause());
+    }
+
+}

--- a/mariaDB4j-core/src/main/java/ch/vorburger/mariadb4j/MariaDB4jService.java
+++ b/mariaDB4j-core/src/main/java/ch/vorburger/mariadb4j/MariaDB4jService.java
@@ -60,15 +60,29 @@ public class MariaDB4jService {
         return configBuilder;
     }
 
-    @PostConstruct
-    // note this is from javax.annotation, not a Spring Framework dependency
+    @PostConstruct // note this is from javax.annotation, not a Spring Framework dependency
+    public void postConstruct() throws ManagedProcessRuntimeException {
+        try {
+            start();
+        } catch (ManagedProcessException e) {
+            throw new ManagedProcessRuntimeException(e);
+        }
+    }
+
     public void start() throws ManagedProcessException {
         db = DB.newEmbeddedDB(getConfiguration().build());
         db.start();
     }
 
-    @PreDestroy
-    // note this is from javax.annotation, not a Spring Framework dependency
+    @PreDestroy // note this is from javax.annotation, not a Spring Framework dependency
+    public void preDestroy() throws ManagedProcessRuntimeException {
+        try {
+            stop();
+        } catch (ManagedProcessException e) {
+            throw new ManagedProcessRuntimeException(e);
+        }
+    }
+
     public void stop() throws ManagedProcessException {
         if (!isRunning())
             return;

